torch/autograd/_functions/reduce.py

```diff
@@ -71,7 +71,10 @@ def safe_zeros_backward(inp, dim):
             exclusive_normal = exclusive_normal_nocp.cumprod(dim)
 
             def reverse_dim(var, dim):
-                return var.index_select(dim, Variable(torch.arange(var.size(dim) - 1, -1, -1)).long())
+                index = Variable(torch.arange(var.size(dim) - 1, -1, -1)).long()
+                if var.is_cuda:
+                    index = index.cuda()
```

Comment:
I believe that when you do `tensor.new()`, you have a new tensor created of the same type and device as `tensor`, so it should work just fine I think. Another (more verbose) option would be to use `with torch.cuda.device_of(tensor):`.

PS: there is a typo in your example, as `var` is a Tensor and the snippet doesn't work.